rotkehlchen/chain/solana/modules/system/decoder.py

```diff
@@ -0,0 +1,71 @@
+
+from rotkehlchen.chain.evm.decoding.constants import OUTGOING_EVENT_TYPES
+from rotkehlchen.chain.solana.decoding.interfaces import SolanaDecoderInterface
+from rotkehlchen.chain.solana.decoding.structures import (
+    DEFAULT_SOLANA_DECODING_OUTPUT,
+    SolanaDecoderContext,
+    SolanaDecodingOutput,
+)
+from rotkehlchen.chain.solana.utils import lamports_to_sol
+from rotkehlchen.constants.assets import A_SOL
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import SolanaAddress
+
+from .constants import SYSTEM_PROGRAM, TRANSFER_DELIMITER
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.decoding.types import CounterpartyDetails
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class SystemDecoder(SolanaDecoderInterface):
+    """Decoder for the System program."""
+
+    def decode_native_transfer(self, context: SolanaDecoderContext) -> SolanaDecodingOutput:
+        """Decode native SOL transfers."""
+        if context.instruction.data[:4] != TRANSFER_DELIMITER:
+            return DEFAULT_SOLANA_DECODING_OUTPUT
+
+        if (account_len := len(context.instruction.accounts)) < 2:
+            log.error(
+                f'Encountered a native transfer instruction in {context.transaction} '
+                f'with {account_len} accounts. Expected 2. Skipping.',
+            )
+            return DEFAULT_SOLANA_DECODING_OUTPUT
+
+        if (direction_result := self.base.decode_direction(
+            from_address=context.instruction.accounts[0],
+            to_address=context.instruction.accounts[1],
+        )) is None:
+            return DEFAULT_SOLANA_DECODING_OUTPUT
+
+        if (raw_amount := int.from_bytes(context.instruction.data[4:12], byteorder='little')) == 0:
+            return DEFAULT_SOLANA_DECODING_OUTPUT
+
+        amount = lamports_to_sol(raw_amount)
+        event_type, event_subtype, location_label, address, counterparty, verb = direction_result
+        counterparty_or_address = counterparty or address
+        preposition = 'to' if event_type in OUTGOING_EVENT_TYPES else 'from'
+        return SolanaDecodingOutput(events=[self.base.make_event_next_index(
+            tx_hash=context.transaction.signature,
+            timestamp=context.transaction.block_time,
+            event_type=event_type,
+            event_subtype=event_subtype,
+            asset=A_SOL,
+            amount=amount,
+            location_label=location_label,
+            notes=f'{verb} {amount} SOL {preposition} {counterparty_or_address}',
+            counterparty=counterparty,
+            address=address,
+        )])
+
+    def addresses_to_decoders(self) -> dict[SolanaAddress, tuple[Any, ...]]:
+        return {SYSTEM_PROGRAM: (self.decode_native_transfer,)}
+
+    @staticmethod
+    def counterparties() -> tuple['CounterpartyDetails', ...]:
+        return ()  # The system program does native solana stuff and doesn't have a counterparty
```

Comment:
isn't this the more reason why it should probably be in the `SolanaTransactionDecoder` class like evm? e.g. `_maybe_decode_simple_transactions`.

this will probably prevent the need to do `pre_address_mappings`